neqo-transport/benches/transfer.rs

```diff
@@ -27,45 +27,69 @@ const TRANSFER_AMOUNT: usize = 1 << 22; // 4Mbyte
+                    ConnectionParameters::default()
+                        .pmtud(true)
+                        .pacing(pacing)
+                        .mlkem(false),
+                    boxed![ReachState::new(State::Confirmed)],
+                    boxed![ReceiveData::new(TRANSFER_AMOUNT)]
+                ),
+                TailDrop::dsl_downlink(),
+                RandomDelay::new(ZERO..JITTER),
+            ];
+            let mut sim = Simulator::new(label, nodes);
+            if let Some(seed) = &seed {
+                sim.seed_str(seed);
+            }
+            sim.setup()
+        };
+        let routine = |sim: ReadySimulator| {
+            black_box(sim.run());
+        };
+
+        let mut group = c.benchmark_group(format!("transfer/pacing-{pacing}/{label}"));
         group.noise_threshold(0.03);
-        group.bench_function(label, |b| {
+
+        // Benchmark with wallclock time, i.e. measure the compute efficiency.
+        group.bench_function("wallclock-time", |b| {
             b.iter_batched(
-                || {
-                    let nodes = boxed![
-                        Node::new_client(
-                            ConnectionParameters::default()
-                                .pmtud(true)
-                                .pacing(pacing)
-                                .mlkem(false),
-                            boxed![ReachState::new(State::Confirmed)],
-                            boxed![SendData::new(TRANSFER_AMOUNT)]
-                        ),
-                        TailDrop::dsl_uplink(),
-                        RandomDelay::new(ZERO..JITTER),
-                        Node::new_server(
-                            ConnectionParameters::default()
-                                .pmtud(true)
-                                .pacing(pacing)
-                                .mlkem(false),
-                            boxed![ReachState::new(State::Confirmed)],
-                            boxed![ReceiveData::new(TRANSFER_AMOUNT)]
-                        ),
-                        TailDrop::dsl_downlink(),
-                        RandomDelay::new(ZERO..JITTER),
-                    ];
-                    let mut sim = Simulator::new(label, nodes);
-                    if let Some(seed) = &seed {
-                        sim.seed_str(seed);
-                    }
-                    sim.setup()
+                setup,
+                |s| {
+                    routine(s);
                 },
-                black_box(ReadySimulator::run),
                 SmallInput,
             );
         });
+
+        // Benchmark with simulated time, i.e. measure the network protocol
+        // efficiency.
+        //
+        // Note: Given that this is using simulated time, we can measure actual
+        // throughput.
+        group.throughput(criterion::Throughput::Bytes(TRANSFER_AMOUNT as u64));
+        group.bench_function("simulated-time", |b| {
+            b.iter_custom(|iters| {
```

Comment:
How does this work out in terms of performance?  Usually, the test infrastructure runs the function a few times to see how long it takes.  Does it use the wall clock for that process, or the time that this function reports?  The wall clock would be fine, but if it uses the time this function reports, that will give some very bad results.

Reply:
Yep, they did the sensible thing.  That's not always something you can rely on.